README


This archive contains the distribution Fey-ORM,
version 0.46:

  A Fey-based ORM (If you like SQL, you might like Fey::ORM)

This software is copyright (c) 2011 by Dave Rolsky.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
0.46   2013-11-28

- Removed used of deprecated Class::MOP::load_class().


0.45   2013-02-22

- Fix test failure from hash randomization changes in 5.17.x Perl.


0.44   2012-05-18

- Work around breakage introduced in MooseX::Types 0.32.


0.43   2011-02-26

- This release provides forward compatibility with Moose 1.99+. It will still
  work with Moose 1.23 as well.


0.42   2011-02-26

- The most recent release of MooseX::StrictConstructor did not work with this
  module. This has been fixed, and Fey::ORM now requires the latest
  MooseX::StrictConstructor.


0.41   2011-02-25

- A small internal refactoring on Fey::Object::Iterator::FromSelect to make it
  easier to override some of its behavior.


0.40   2011-01-25

- A stray extends statement left in from a refactoring in 0.39 caused test
  failures unless a previous version of Fey::ORM was already installed.


0.39   2011-01-25

- When you explicitly provide a Fey::FK object for a has_one or has_many
  relationship, Fey::ORM will now invert the direction of that relationship if
  needed to generate the proper SQL. In the past, this was done as needed if
  the Fey::FK _wasn't_ explicitly provided. Now it is done regardless.


0.38   2011-01-07

- Allow the use of SQL queries as insert & update values.


0.37   2010-11-30

- The caching added in 0.34 had some subtle heisenbugs because it used object
  refaddrs as cache keys. Now we use Object::ID, which is slower but will not
  cause false positive cache hits.


0.36   2010-10-22

- Some packaging tweaks so that the Fey::ORM::Manual::Intro docs show up on
  search.cpan.


0.35   2010-10-05

- Changes to work with Moose 1.15.


0.34   2010-09-09

- Added caching for commonly used stringified SQL queries. This speeds up
  common operations like selecting a single object.


0.33   2010-06-04

- When loading an object's data from the dbms, Fey::ORM now always reloads the
  key. This should impose a minimal cost, and is important in some corner
  cases, like Postgres's citext column, which does a case-insensitive
  match. In that case, simply using the user-provided data can cause the
  object to end up with the wrong value for the column in question.


0.32   2010-03-06

- Added a new sugar function to Fey::ORM::Table - query(). This allows you to
  declare a method based on a select. This is preferable to an attribute in
  many cases, where you don't want to cache the return value of the SELECT in
  the object.


0.31   2010-02-12

- Fey::ORM::Table tried to load a module that no longer exists with the newer
  versions of MooseX::ClassAttribute.


0.30   2009-10-25

- The Fey::Object::Iterator::FromSelect class used to expect that the number
  of elements in the select object's select clause matched the number of items
  actually returned by the select in each row. However, if you used a Term to
  do a "SELECT *" this assumption was broken. The assumption meant that even
  providing an explicit attribute map in this case did not work.

- As a corrolary, Fey::Object::Iterator::FromSelect used to attempt to merge
  an explicit attribute map with what was implied by any columns in the select
  object's select clause. This was never actually documented, and has been
  removed.

  Now, providing an explicit map means you must provide a mapping for every
  select clause element you want to pass to an object.

- Added Fey::Object::Iterator::FromSelect->raw_row(). This method lets you get
  the raw data for a row after calling ->next(). This is handy if the select
  contained data that does not fit into a particular class.

0.29   2009-10-10

- Allow any classes to be passed to an Iterator's constructor. However, unless
  all the classes have a Table() method, you cannot get hashes back from the
  iterator (next_as_hash(), etc.).


0.28   2009-09-15

- When a Fey::ORM::Table using class's constructor does not return an object,
  the error message is no longer set in $@. Instead, you can call
  MyClass->ConstructorError() to get the error. The value of $@ is preserved
  in the constructor as whatever it was before the constructor was called.

- Added a $VERSION to every .pm file that matches Fey::ORM.

- Doc fixes - RT #48484 (Terrence Brannon) and #49595 (Guoying Qi)


0.27   2009-08-18

- Changed all code that checked for a Fey::SQL::Select object to instead check
  that the object does the Fey;:Role::SQL::ReturnsData role. This lets you
  pass a union to an object iterator, among other things.


0.26   2009-07-03

- Changes to work with the latest Moose (0.86).


0.25   2009-06-21

- Changes to work with the latest Moose (0.82).


0.24   2009-04-20

- Shut up a warning from the latest Moose that it will not inline a
  constructor for Fey::Object::Table.


0.23   2009-04-07

- Changes to work with the latest Moose (0.73_01+).


0.22   2009-03-23

- If a table consisted of _just_ its key, then trying to call new()
  for the associated class blew up, but they could be created via an
  iterator. Test case provided by Andrew Baumhauer.


0.21   2009-02-25

* This release contains several backwards incompatible changes.

* Fey::Object::Iterator has been renamed to
  Fey::Object::Iterator::FromSelect. Fey::Object::Iterator::Caching is
  now has Fey::Object::Iterator::FromSelect::Caching. This was done to
  allow for a new class, Fey::Object::Iterator::FromArray. There is
  now a Fey::ORM::Role::Iterator role which all of these classes
  do. Based on a patch from Bryan Deeney and Will Weaver.

* The iterator methods all and all_as_hashes now really return _all_
  of the data, which means they will reset the iterator if needed.

- Added remaining and remaining_as_hashes methods to iterators. These
  do what all and all_as_hashes did previously. Thanks 

- The transform feature now supports passing delegation for the
  inflated object. See Fey::ORM::Table for details.


0.20   2009-02-06

- Use MooseX::Params::Validate instead of the no-longer-in-Fey
  Fey::Validate.


0.19   2009-01-18


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A Fey-based ORM (If you like SQL, you might like Fey::ORM)",
  "AUTHOR" => "Dave Rolsky <autarch\@urth.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Fey-ORM",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Fey::ORM",
  "PREREQ_PM" => {
    "Class::Load" => 0,
    "Class::MOP::Method::Generated" => 0,
    "Devel::GlobalDestruction" => 0,
    "Exception::Class" => 0,
    "Fey" => "0.39",
    "Fey::DBIManager" => "0.07",
    "Fey::Exceptions" => 0,
    "Fey::Literal::Function" => 0,
    "Fey::Placeholder" => 0,
    "Fey::SQL" => 0,
    "Fey::Table" => 0,
    "List::AllUtils" => 0,
    "List::Util" => 0,
    "Moose" => "1.15",
    "Moose::Exporter" => 0,
    "Moose::Meta::Attribute" => 0,
    "Moose::Meta::Class" => 0,
    "Moose::Meta::Method" => 0,
    "Moose::Meta::Method::Constructor" => 0,
    "Moose::Role" => 0,
    "Moose::Util::MetaRole" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "MooseX::ClassAttribute" => "0.24",
    "MooseX::Params::Validate" => 0,
    "MooseX::SemiAffordanceAccessor" => 0,
    "MooseX::StrictConstructor" => "0.13",
    "MooseX::Types" => 0,
    "MooseX::Types::Combine" => 0,
    "MooseX::Types::Moose" => 0,
    "Object::ID" => 0,
    "Scalar::Util" => 0,
    "Sub::Exporter" => 0,
    "Try::Tiny" => 0,
    "base" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Exporter" => 0,
    "Fey::FakeDBI" => 0,
    "Fey::Literal::String" => 0,
    "Fey::Test" => "0.05",
    "Test::Fatal" => 0,
    "Test::More" => "0.88",
    "lib" => 0
  },
  "VERSION" => "0.46",
  "test" => {
    "TESTS" => "t/*.t t/Class/*.t t/Object/*.t t/Object/Iterator/*.t t/Object/Iterator/FromSelect/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



